### ISSUE TITLE
tap to bring to front?

### ISSUE THREAD

**Issue by vixalien** [Moon-0xff/gnome-mpris-label#44]
I don't know how sp-tray does it, but when you click on the indicator, it jumps to the spotify app and when you click it again, it jumps back to where you were.

**Comment by Moon-0xff** [comment]
Since yesterday's update you can press right-click (default binding) in the label to raise the window, right-clicking again does nothing. So we are already halfway there.

This was a mere oversight, and shouldn't be a difficult change to add (i hope).  
Answering the question of *how sp-tray does it* i think all comes down to this function:
```js
        // many thanks to mheine's implementation
        showSpotify() {
            if (this._spotiWin && this._spotiWin.has_focus()) {
                // spotify is up and focused
                if (this._notSpotify) {
                    // hide spotify and pull up the last active window if possible
                    Main.activateWindow(this._notSpotify);
                }
            } else {
                // spotify is unfocused or the tray icon has never been clicked before
                this._spotiWin = this._notSpotify = null;
                let wins = global.get_window_actors(); // get all open windows
                for (let win of wins) {
                    if (typeof win.get_meta_window === "function") {
                        if (win.get_meta_window().get_wm_class() === "Spotify") {
                            this._spotiWin = win.get_meta_window(); // mark the spotify window
                        } else if (win.get_meta_window().has_focus()) {
                            this._notSpotify = win.get_meta_window(); // mark the window that was active when the button was pressed
                        }

                        if (this._spotiWin && this._notSpotify) {
                            break;
                        }
                    }
                }
                Main.activateWindow(this._spotiWin); // pull up the spotify window
            }
        }
```

**Comment by vixalien** [comment]
oh that's good to hear. let me re-login to checkout the new features

**Comment by Batwam** [comment]
@Moon-0xff Note that the current implementation in #29 opens the player but does not put the previously focussed window back on with second click.

I had the plan of hiding the window on second click (actually had a TODO note in the code at some point) as it felt like a natural extension of the action until I realised that we don't really know what the user is doing between the first and second click. For waht we know multiple hours could have passed so we could be showing the wrong app. Our implementation is also a little bit different than sp-tray as sp-tray only works with Spotify while mpris-label works with every music player known to man. This means that a user could click once to open Spotify, close it, start Rhythmbox and expect the `Open App` click to open Rhythmbox.

Basically it felt like to much work for something the user can achieve more reliably by typiing alt+tab. Actually, now that I think about it, would there a way to perform the alt+tab action in gjs? this could be a way to implement this easily if we felt it was needed?

**Comment by vixalien** [comment]
my tiny brains thought that the "recent windows" could be got by an API or something. I mean the list of windows a user see when they tap `Alt+Tab`. I thought we could then switch to the next app on that list (provided it isn't the current player)

**Comment by Moon-0xff** [comment]
Yes, we can trigger an alt+tab by interfacing with the `AppSwitcherPopup` on `Main`. That's my theory at least.

And we can guess what's the player window(s) by name like sp-tray does to find the spotify window:  
```js
  let wins = global.get_window_actors(); // get all open windows
                for (let win of wins) {
                    if (typeof win.get_meta_window === "function") {
                        if (win.get_meta_window().get_wm_class() === "Spotify") {
                            this._spotiWin = win.get_meta_window(); // mark the spotify window
```

**Comment by vixalien** [comment]
hmm this is interesting

**Comment by Batwam** [comment]
first pass code implementaiton in the PR above

**Comment by Batwam** [comment]
some example of use of AltTab here if this makes sense to you: https://github.com/G-dH/advanced-alttab-window-switcher/blob/main/extension.js

**Comment by Batwam** [comment]
thanks @Moon-0xff for the rapid review and insightful suggestions always. As this is now merged, should we push this to the general public so everyone can enjoy the benefits of @vixalien's suggestion as a new release? It feels like merging the controls and filters might delay this a bit too long. 

Note that while we are at it, it might be worth including #45 too as it's a pretty simple one.

**Comment by Moon-0xff** [comment]
I share the feeling that #31 and #39 are far from being merged, but this is a rather small feature to solely justify an update.

**Comment by Moon-0xff** [comment]
New update includes #46
Thanks @vixalien for raising this issue, and of course, @Batwam for the PR.

**Comment by vixalien** [comment]
Maybe I should probably create a new issue, but there are a few issues with the current implementation.

For example, when I'm playing with Firefox, clicking the label in the top bar causes Firefox to just open, instead of taking me to the currently playing tab.

Clicking on the MPRIS notification instead, takes me correctly to the playing tab instead of just opening Firefox.

**Comment by Batwam** [comment]
yeah, please create a new issue as this one is closed.
